### PR TITLE
ACS-3229: Rules v1 REST API - Get rule definition

### PR DESCRIPTION
**Pull request by krdabrowski**
`GET /nodes/{folderNodeId}/rule-sets/{ruleSetId}/rules/{ruleId}` implementation.

Initially return minimal id & name (similar to previous listing rules). Note: the triggers, conditions, actions and other properties will follow-on in subsequent PR.

### CODE REVIEW

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java [hunk 1]
```diff
@@ -93,37 +95,80 @@ public void setRuleService(RuleService ruleService)
...
+     * @param ruleSetId - rule set node ID
+     * @param associatedFolderNodeRef - folder node ref to check the association
+     * @return rule set node reference or null if given rule set ID is the default one
+     * @throws InvalidArgumentException in case of not matching associated folder node
      */
+    private NodeRef validateRuleSetNode(final String ruleSetId, final NodeRef associatedFolderNodeRef)
+    {
+        if (RuleSet.isDefaultId(ruleSetId))
+        {
+            return null;
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:129]: Nice catch, didn't think about that - I will fix it.

**Review by tpage-alfresco — APPROVED** [review]
LGTM

**Comment by montgolfiere** [comment]
JFYI: one test failure related to rules:

> [ERROR] Tests run: 285, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 366.768 s <<< FAILURE! - in org.alfresco.AppContext01TestSuite
> [ERROR] testIsRuleAssociatedWithRuleSet(org.alfresco.repo.rule.RuleLinkTest)  Time elapsed: 0.086 s  <<< FAILURE!
> junit.framework.AssertionFailedError
>	at org.alfresco.repo.rule.RuleLinkTest.testIsRuleAssociatedWithRuleSet(RuleLinkTest.java:403)

plus one other that might be unrelated / intermittent ?

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:133]: After rethinking I've added a comment, as the value won't be ever used.

**Review by tpage-alfresco — CHANGES_REQUESTED** [review]
Please can you remove all the extra changes around Actions/Conditions from the PR. We can review those separately.

File: data-model/src/main/java/org/alfresco/service/namespace/NamespaceService.java [hunk 2]
```diff
@@ -146,7 +146,7 @@ public interface NamespaceService extends NamespacePrefixResolver
 
     /** Email Server Application Model Prefix */
     static final String EMAILSERVER_MODEL_PREFIX = "emailserver";
-    
+
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply data-model/src/main/java/org/alfresco/service/namespace/NamespaceService.java:149]: Yeah... unfortunately IDE is stubborn and wont let my fix it (add there those whitespaces)

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java [hunk 3]
```diff
@@ -93,37 +95,80 @@ public void setRuleService(RuleService ruleService)
         this.ruleService = ruleService;
     }
 
-    private NodeRef validateNode(final String nodeId, final QName namespaceType)
+    /**
+     * Validates if folder node exists and user have permission to read from it.
+     *
+     * @param folderNodeId - folder node ID
+     * @return folder node reference
+     * @throws InvalidArgumentException if node is not of an expected type
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:105]: Added to the javadoc

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java [hunk 4]
```diff
@@ -93,37 +95,80 @@ public void setRuleService(RuleService ruleService)
...
+     * Validates if rule set ID is default, node exists and associated folder node matches.
      *
-     * @param nodeId - node ID
-     * @param expectedType - expected type
-     * @param expectedTypeName - expected type local name
-     * @return node reference
+     * @param ruleSetId - rule set node ID
+     * @param associatedFolderNodeRef - folder node ref to check the association
+     * @return rule set node reference or null if given rule set ID is the default one
+     * @throws InvalidArgumentException in case of not matching associated folder node
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:126]: In fact this method doesn't throw PermissionDeniedException. Read permission is only checked for folder node.

File: remote-api/src/test/java/org/alfresco/rest/framework/tests/core/ParamsExtender.java [hunk 5]
```diff
@@ -43,29 +43,33 @@
 public class ParamsExtender extends Params
 {
 
-    private ParamsExtender(String entityId, String relationshipId, Object passedIn, InputStream stream, String addressedProperty, RecognizedParams recognizedParams)
+    private ParamsExtender(String entityId, String relationshipId, String relationship2Id, Object passedIn, InputStream stream, String addressedProperty, RecognizedParams recognizedParams)
     {
-        super(null,entityId, relationshipId, null, passedIn, stream, addressedProperty, recognizedParams, null, mock(WebScriptRequest.class));
+        super(null,entityId, relationshipId, relationship2Id, passedIn, stream, addressedProperty, recognizedParams, null, mock(WebScriptRequest.class));
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/test/java/org/alfresco/rest/framework/tests/core/ParamsExtender.java:48]: In tests I needed to specify relationship2Id (ruleId) - this is the only reason of this fix.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java [hunk 6]
```diff
@@ -93,37 +95,80 @@ public void setRuleService(RuleService ruleService)
...
      */
+    private NodeRef validateRuleSetNode(final String ruleSetId, final NodeRef associatedFolderNodeRef)
+    {
+        if (RuleSet.isDefaultId(ruleSetId))
+        {
+            return null;
+        }
+
+        final NodeRef ruleSetNodeRef = validateNode(ruleSetId, ContentModel.TYPE_SYSTEM_FOLDER, "rule set");
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java]: I've replaced string literal with a constant. Unfortunately I didn't found needed string in any set/enum.

**Review by tpage-alfresco — APPROVED** [review]
LGTM!

**Review by montgolfiere — APPROVED** [review]
LGTM (subject to usual green build)

As agreed, follow-on PRs will add the the triggers, conditions, actions and other properties (for "CRU") :-)

**Pull request merged**